### PR TITLE
Rustdoc: add crate items to the search-index

### LINKED ISSUE #14077: rustdoc search doesn't find crate names
E.g. [searching for `libc`](http://static.rust-lang.org/doc/master/std/index.html?search=libc) doesn't show the `libc` crate.

### PR DESCRIPTION
**Pull request by mitaa**
This allows to search for crates in documentation and simplifies the json serialization of the search-index.

fixes #14077

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -290,23 +290,36 @@ struct IndexItem {
...
-        // but just to be safe we write `null` in that case.
-        match self.name {
-            Some(ref n) => write!(f, "{{\"name\":\"{}\"}}", n),
-            None => write!(f, "null")
-        }
+impl ToJson for Type {
+    fn to_json(&self) -> Json {
+        let mut data = BTreeMap::new();
+        data.insert("name".to_owned(), self.name.to_json());
+        Json::Object(data)
```
Comment (alexcrichton) [inline src/librustdoc/html/render.rs]: This seems different than it was before? Just double-checking that it was accounted for on the search index side of things

**Comment by alexcrichton** [comment]
Sorry this took awhile to get around to, but thanks @mitaa! I like the cleanup going on here, definitely improves readability.

I've got just one minor question about how the search index looks like it's changed, but other than that looks good to me.

**Comment by mitaa** [comment]
You're right, I've changed that part back to the old behaviour - the search index shouldn't have changed but it's probably safer this way. (like the comment said the `null` case shouldn't normally happen)

**Pull request merged**